### ISSUE TITLE
_bulk_get uses compression inefficiently

### ISSUE THREAD

**Issue by snej** [couchbase/sync_gateway#1163]
A customer is reporting that the initial pull to a new database is using several times more bandwidth than their old protocol was, due to inefficient compression. From Perry Krug's email:

> It appears that their legacy system compresses the entire dataset and sends it to the mobile device in one-shot.  That results in around 18k of data transferred over the wire.  However, our implementation is taking nearly 1.25MB to transfer the exact same data.  This also leads us to have ~5s of delay for the data to load whereas their legacy implementation was sub-1s.
> 
> Looking deep into the packet captures I can see that the data _is_ being compressed. However, it is being compressed on a per-doc basis instead of the whole HTTP response and so I believe this is causing us to send a much larger amount of data over the wire.

**Comment by snej** [comment]
The `_bulk_get` response from the server is a MIME multipart body of individual gzipped documents, rather than a gzipped body of uncompressed documents. That is, it does not use HTTP response compression (`Content-Encoding: gzip`) on the entire response, but instead on the individual nested MIME parts corresponding to JSON documents.

This is less efficient because it doesn't exploit the commonality between documents.

This could be changed on the sending end (Sync Gateway) without affecting Couchbase Lite. On the client side, the added HTTP gzip Content-Encoding will be transparently decoded, and the code that parses the individual MIME bodies looks for a Content-Encoding header before decompressing, so if those aren’t there it’ll treat them as uncompressed.

**Comment by snej** [comment]
I spent a little while looking into the Sync Gateway code for response compression, and found a few things:
- Currently, we only enable HTTP compression for responses whose Content-Type is JSON or text. We’d have to extend this to multipart. [encoded_response_writer.go:67]
- The individual per-document compression in a `_bulk_get` response is enabled only if the client request includes an "X-Accept-Part-Encoding” header, so it can be disabled on the client side by commenting out the line that adds that header. [bulk_api.go:300]

Also, on further thought, I think it may not be good in general to use HTTP-level compression of `_bulk_get` responses, because those responses interleave JSON and attachments. In some cases the bulk of the response will be attachment data, and such data is likely to be already compressed (JPEGs, MP3s, video…), so running it through gzip will waste CPU on the server side. It will also make the data slightly larger and use a bit of CPU on the client, but those effects should be minor.

The real solution for this would be to change the CBL replicator to download only the document bodies with `_bulk_get`, and then fetch the attachments as individual GETs. (This ties in nicely with lazy attachment downloading, too.) This isn’t a big change, but it’s larger-scale than what I was thinking yesterday. I've filed it in the iOS repo as couchbase/couchbase-lite-ios#923.
